File: Day99/Part4/chaos_world.py
import sys

def solve():
    input_data = sys.stdin.read().split()
    if not input_data:
        return
    iterator = iter(input_data)
    try:
        N = int(next(iterator))
        # Assuming functional graph layout input
        # No M if outdegree=1.
        # But if problem format is standard graph list...
        # Wait, if world changes, maybe functional graph changes?
        # Re-reading: "World berubah tiap langkah".
        # This implies u -> v depends on t.
        # Maybe F_t(u).
        # Assuming simple static functional graph for now based on "day99" patterns.
        # If input has M edges: u, v. And we pick?
        # "Strategi". Implies choice.
        # If random choice? No, "Strategi pasti berhasil". 
        # This is Synchronizing Word problem (DFA).
        # We assume input is DFA: N states, K symbols. Transitions u,c -> v.
        # Task: Find sequence of symbols that maps {1..N} -> {Target}.
        # Sync word existence is PSPACE? Or just BFS in power set?
        # Power set BFS.
        
        M_inputs = int(next(iterator)) # Number of symbols/moves available
        Target = int(next(iterator))
        
        adj = {} # (u, symbol) -> v
        # Read transitions: u, symbol, v
        # Usually N*M lines? Or sparse.
        
        # Let's assume input format:
        # N, Symbols, Target.
        # Then N*Symbols lines? Or M edges.
        
        # Heuristic reader
        num_edges = int(next(iterator)) # Explicit edge count
        for _ in range(num_edges):
            u = int(next(iterator))
            sym = int(next(iterator))
            v = int(next(iterator))
            adj[(u, sym)] = v
            
    except StopIteration:
        return

    # BFS on Subset State
    # Start State: All nodes {1..N}
    # Goal: {Target}
    # Transitions: For each symbol s, next_state = { adj[u][s] for u in current }
    
    initial_set = frozenset(range(1, N+1))
    queue = [initial_set]
    visited = {initial_set}
    
    import collections
    q = collections.deque([initial_set])
    
    while q:
        curr = q.popleft()
        
        if len(curr) == 1 and Target in curr:
            print("YES")
            return
            
        for sym in range(1, M_inputs + 1): # Assume symbols 1..M
            next_set = set()
            valid_move = True
            for u in curr:
                if (u, sym) in adj:
                    next_set.add(adj[(u, sym)])
                else:
                    # Deterministic? If edge missing, maybe self loop or invalid?
                    # Assume self loop or dead end (empty).
                    # If dead end, set shrinks (good).
                    # If undefined, assume stays?
                    # Let's assume strict input: all defined or implicit dead.
                    pass 
            
            if next_set:
                frozen = frozenset(next_set)
                if frozen not in visited:
                    visited.add(frozen)
                    q.append(frozen)
    
    print("NO")

if __name__ == "__main__":
    solve()

File: Day99/Part4/chaos_world.cpp
#include <iostream>
#include <vector>
#include <set>
#include <algorithm>

using namespace std;

// Problem 4: Procedural World With Deterministic Chaos
// Rule: Deterministic evolution.
// Unknown Start.
// Goal: Reach Target surely.
// Approach: Maintain set of ALL possible current nodes.
// Transitions: S_new = { rule(u) for u in S }.
// If S_new == {Target}, YES.
// If S_new enters cycle without collapsing to Target, NO (or keep searching if loop matches).
// If steps > Limit, NO.

int main() {
    int N, M, Target;
    if (!(cin >> N >> M >> Target)) return 0;
    
    // Graph representation
    // Edges u->v at step t? 
    // "Rule deterministik". May imply single successor?
    // "World berubah tiap langkah". 
    // Maybe Edge(u, t%P) -> v? Or Input gives transition table for T steps?
    // Let's assume standard automata: N nodes, M input symbols?
    // "Kamu tahu rule". If rule is just time-based, no input.
    // If no input, we just simulate.
    // Input: Edges. Maybe simple directed graph where out-degree=1?
    // Or time dependent? 
    // Assume functional graph u->v.
    // Start with Set = {1..N}.
    // Step repeatedly.
    // Since graph is finite and functional, we enter cycles.
    // Check if the set of possible locations becomes {Target}.
    
    // Input reading: Assume functional graph u->next(u). 
    // If multiple edges map time?
    // Let's assume u -> v.
    
    vector<int> next_node(N + 1);
    for(int i=1; i<=N; ++i) {
        int v; cin >> v; // Input next node for i
        next_node[i] = v;
    }
    
    set<int> current_possible;
    for(int i=1; i<=N; ++i) current_possible.insert(i);
    
    for(int step=0; step < 2 * N + 5; ++step) { // Run enough to enter cycles
        if (current_possible.size() == 1 && *current_possible.begin() == Target) {
            cout << "YES" << endl;
            return 0;
        }
        
        set<int> next_possible;
        for(int u : current_possible) {
            next_possible.insert(next_node[u]);
        }
        current_possible = next_possible;
    }
    
    cout << "NO" << endl;
    return 0;
}
